### PR TITLE
[API-582] Implement external smart client discovery

### PR DESCRIPTION
**Pull request by mdumandag**
This PR implements the client-side support for the external smart
client discovery. With this PR, the Python client will be able
to use public addresses reported by the cluster members even
if the client and the cluster are in different networks.

To use it, one has to enable `use_public_addresses` option by setting
it to `True`, and passing at least one public address of the cluster
members to the `cluster_members` config. Client will initiate the
cluster connection using the provided public address, and will
discover the rest of the cluster using the public addresses reported
by them.

Protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/394

### CODE REVIEW

**Comment by mdumandag** [comment]
Ali did the tests on Kubernetes on GKE with this branch, and the tests are passed.

Here is the link to his verification: https://github.com/alisengul53/GHActionTestRepo/runs/3542770380?check_suite_focus=true

File: hazelcast/cluster.py [hunk 1]
```diff
@@ -284,6 +287,17 @@ def _members_string(members):
     def _create_snapshot(version, member_infos):
         new_members = OrderedDict()
         for member_info in member_infos:
+            address_map = member_info.address_map
+            if address_map:
+                address = address_map.get(
+                    _CLIENT_ENDPOINT_QUALIFIER,
+                    address_map.get(_MEMBER_ENDPOINT_QUALIFIER, member_info.address),
```
Comment (srknzl) [inline hazelcast/cluster.py:294]: Isn't this second default value `member_info.address` extra compared to java implementation? In Java if address_map.get(_MEMBER_ENDPOINT_QUALIFIER) will return null, address of member will be null. 

```java
memberBuilder = new MemberImpl.Builder(addressMap)
                        .address(addressMap.getOrDefault(CLIENT, addressMap.get(MEMBER)));
```

File: hazelcast/cluster.py [hunk 2]
```diff
@@ -284,6 +287,17 @@ def _members_string(members):
...
+                )
+                member_info.address = address
+            else:
+                # It might be None on 4.0 servers.
+                member_info.address_map = {}
```
Comment (srknzl) [inline hazelcast/cluster.py:299]: In java if address map is null, memberBuilder's build() sets it to a map that includes single entry: member endpoint qualifier -> address.

Why we set it to an empty map?

File: hazelcast/proxy/reliable_topic.py [hunk 3]
```diff
@@ -240,7 +240,13 @@ def _handle_next_batch(self, future):
...
                     if message.publisher_address:
                         member = MemberInfo(
-                            message.publisher_address, None, None, False, _UNKNOWN_MEMBER_VERSION
+                            message.publisher_address,
+                            None,
+                            None,
+                            False,
+                            _UNKNOWN_MEMBER_VERSION,
+                            None,
+                            {},
```
Comment (srknzl) [inline hazelcast/proxy/reliable_topic.py:249]: In java, address map of this memberInfo is set to a map that includes single entry: member endpoint qualifier -> address

File: hazelcast/connection.py [hunk 4]
```diff
@@ -361,6 +371,16 @@ def _translate(self, address):
 
         return translated
 
+    def _translate_member_address(self, member):
+        if self._use_public_addresses:
+            public_address = member.address_map.get(_CLIENT_PUBLIC_ENDPOINT_QUALIFIER, None)
```
Comment (srknzl) [inline hazelcast/connection.py:376]: we don't need to specify default None

**Review by srknzl — COMMENTED** [review]
I started to review tests. Just wanted to send the reviews so far..

File: hazelcast/config.py [hunk 5]
```diff
@@ -1335,6 +1337,19 @@ def token_provider(self, token_provider):
             raise TypeError("token_provider must be an object with a token method")
         self._token_provider = token_provider
 
+    @property
+    def use_public_addresses(self):
+        # type: () -> bool
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/config.py:1342]: You can omit the type of the `self`, so they are in fact equal. 

We started to add type information for some methods as best effort basis to make our life's better :smile: Pycharm can understand that information and help us in some occasions. You are right, we should add those to all methods but that's a job for later

File: hazelcast/core.py [hunk 6]
```diff
@@ -326,7 +415,8 @@ class HazelcastJsonValue(object):
     """
 
     def __init__(self, value):
-        util.check_not_none(value, "JSON string or the object cannot be None.")
+        if value is None:
+            raise AssertionError("JSON string or the object cannot be None.")
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/core.py:419]: Due to Cyclic imports. I had to remove the import to `hazelcast.util`

We need to spend some time enhancing the module structure of the client, possibly in 5.0 so hopefully, we will hit such problems less

File: hazelcast/core.py [hunk 7]
```diff
@@ -374,6 +464,9 @@ def __init__(self, major, minor, patch):
         self.minor = minor
         self.patch = patch
 
+    def __repr__(self):
+        return "MemberVersion(major=%s, minor=%s, patch=%s)" % (self.major, self.minor, self.patch)
+
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/core.py:469]: Yes

File: hazelcast/protocol/codec/custom/endpoint_qualifier_codec.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 from hazelcast.protocol.builtin import FixSizedTypesCodec, CodecUtil
 from hazelcast.serialization.bits import *
 from hazelcast.protocol.client_message import END_FRAME_BUF, END_FINAL_FRAME_BUF, SIZE_OF_FRAME_LENGTH_AND_FLAGS, create_initial_buffer_custom
-from hazelcast.protocol import EndpointQualifier
+from hazelcast.core import EndpointQualifier
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/protocol/codec/custom/endpoint_qualifier_codec.py:4]: Yes, I will send the PR tomorrow

File: hazelcast/core.py [hunk 9]
```diff
@@ -14,12 +13,12 @@ class MemberInfo(object):
     Represents a member in the cluster with its address, uuid, lite member status, attributes and version.
     """
 
-    __slots__ = ("address", "uuid", "attributes", "lite_member", "version")
+    __slots__ = ("address", "uuid", "attributes", "lite_member", "version", "address_map")
 
-    def __init__(self, address, uuid, attributes, lite_member, version, *_):
+    def __init__(self, address, uuid, attributes, lite_member, version, _, address_map):
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/core.py:18]: it is for the is_address_map_exists field, which is unused. So, I renamed it `_`. It is used in the protocol so, I need it here

File: hazelcast/core.py [hunk 10]
```diff
@@ -14,12 +13,12 @@ class MemberInfo(object):
     Represents a member in the cluster with its address, uuid, lite member status, attributes and version.
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/core.py:13]: I couldn't get this comment. Can you rephrase it?

[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/core.py:13]: sorry somehow left like this. I was saying we could add address map to this sentence maybe

File: tests/integration/connection_manager_translate_test.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        conn = conn_manager._get_or_connect_to_member(member).result()
+        self.assertIsNotNone(conn)
+
+    def test_translate_is_not_used_when_member_has_public_client_address_but_option_is_disabled(
+        self,
+    ):
+        self.client = HazelcastClient(
+            cluster_name=self.cluster.id,
+            cluster_connect_timeout=1.0,
+            connection_timeout=1.0,
```
Comment (srknzl) [inline tests/integration/connection_manager_translate_test.py:101]: again these timeouts do not exist on java. Why we needed them?

File: tests/integration/connection_manager_translate_test.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
+import uuid
```
Comment (srknzl) [inline tests/integration/connection_manager_translate_test.py:1]: Related to tests in general. Did you understand what testTranslateFromMemberIsNotUsedWhenAlreadyTranslatedByAddressProvider test does in java side? I am confused with the test and also I see that you don't include it. Was it hard for you to get as well?

**Comment by yuce** [comment]
@mdumandag The Java client property that enables this feature is `hazelcast.discovery.public.ip.enabled`: https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/2689663054/External+Smart+Client+discovery+via+LoadBalancer+TDD I've also used `config.Cluster.Discovery.UsePublicIP` in the Go client. So, would it make sense to use something similar in the Python client as well?

File: tests/integration/connection_manager_translate_test.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+from mock import patch
+
+from hazelcast import HazelcastClient
+from hazelcast.core import Address, MemberInfo, MemberVersion, EndpointQualifier, ProtocolType
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:6]: No, this is not in the backward compatibility test suite, as this checks the protected methods of the connection manager, which can change from release to release

File: tests/integration/connection_manager_translate_test.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+from hazelcast.errors import IllegalStateError, TargetDisconnectedError
+from tests.base import HazelcastTestCase
+
+_UNREACHABLE_ADDRESS = Address("192.168.0.1", 5701)
+_MEMBER_VERSION = MemberVersion(5, 0, 0)
+_CLIENT_PUBLIC_ENDPOINT_QUALIFIER = EndpointQualifier(ProtocolType.CLIENT, "public")
+
+
+class ConnectionManagerTranslateTest(HazelcastTestCase):
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:15]: I don't like to use it when I don't need the client created in the SingleMemberTestCase

File: tests/integration/connection_manager_translate_test.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+from mock import patch
+
+from hazelcast import HazelcastClient
+from hazelcast.core import Address, MemberInfo, MemberVersion, EndpointQualifier, ProtocolType
+from hazelcast.errors import IllegalStateError, TargetDisconnectedError
+from tests.base import HazelcastTestCase
+
+_UNREACHABLE_ADDRESS = Address("192.168.0.1", 5701)
+_MEMBER_VERSION = MemberVersion(5, 0, 0)
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:11]: We don't need to maintain anything here. This is just a bogus data we put to imaginary member instances we create.

File: tests/integration/connection_manager_translate_test.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    def test_translate_is_not_used_when_getting_existing_connection(self):
+        provider = TestAddressProvider(False, self.member.address)
+        with patch.object(
+            HazelcastClient,
+            "_create_address_provider",
+            return_value=provider,
+        ):
+            self.client = HazelcastClient(
+                cluster_name=self.cluster.id,
+                cluster_connect_timeout=1.0,
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:63]: Not really important, as it does not affect the behavior we test. So, I will leave it like this

File: hazelcast/cluster.py (hunk 1 above)
Comment (mdumandag) [reply hazelcast/cluster.py:294]: I don't like the behaviour but anyways, I reverted my addition

File: hazelcast/connection.py (hunk 4 above)
Comment (mdumandag) [reply hazelcast/connection.py:376]: I know, but, I like being explicit here.  If you look at the client codebase, you would see that there is almost always an explicit default value

File: tests/integration/connection_manager_translate_test.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
+import uuid
+
+from mock import patch
+
+from hazelcast import HazelcastClient
+from hazelcast.core import Address, MemberInfo, MemberVersion, EndpointQualifier, ProtocolType
+from hazelcast.errors import IllegalStateError, TargetDisconnectedError
+from tests.base import HazelcastTestCase
+
+_UNREACHABLE_ADDRESS = Address("192.168.0.1", 5701)
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:10]: It is not a public address, it is used in different fields/methods as a just unreachable address so I will leave it like it

File: tests/integration/connection_manager_translate_test.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+        member = MemberInfo(
+            _UNREACHABLE_ADDRESS,
+            uuid.uuid4(),
+            [],
+            False,
+            _MEMBER_VERSION,
+            None,
+            {
+                _CLIENT_PUBLIC_ENDPOINT_QUALIFIER: self.member.address,
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:88]: Not needed for test, so let's ignore it

File: tests/integration/connection_manager_translate_test.py (hunk 12 above)
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:1]: Yes, basically, if you set the property to true and use a custom address provider, we will ignore the logic of the address translation for members altogether. So, the test seems weird to me

File: hazelcast/protocol/codec/custom/endpoint_qualifier_codec.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/protocol/codec/custom/endpoint_qualifier_codec.py:4]: https://github.com/hazelcast/hazelcast-client-protocol/pull/394

File: tests/integration/connection_manager_translate_test.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply tests/integration/connection_manager_translate_test.py:11]: Indeed. Actually it is not used at all in get_or_connect_to_member

File: tests/integration/connection_manager_translate_test.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply tests/integration/connection_manager_translate_test.py:63]: Actually I dont get it why it is used at all. Only testTranslateIsUsed uses a timeout. It can lead to test failures if connecting to the cluster takes more than 1 second. Can't it?

File: tests/integration/connection_manager_translate_test.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply tests/integration/connection_manager_translate_test.py:10]: Java test refers this address as public address? 

```java
// incorrect public address
publicAddress = new Address("192.168.0.1", 5701);
```

**Comment by mdumandag** [comment]
@yuce I changed it to `use_public_ip` to match with others. WDYT?

At first, I used `use_public_addresses` because it sounded better to me but I really don't have any strong opinions about it.

File: tests/integration/connection_manager_translate_test.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py]: This timeout is only considered when the first connection to the cluster fails. So, it can take longer than that.

Not really important, but let me remove them

File: tests/integration/connection_manager_translate_test.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply tests/integration/connection_manager_translate_test.py:10]: It is just an unreachable address. It being public or not is not important IMHO.

File: hazelcast/client.py [hunk 19]
```diff
@@ -332,6 +332,9 @@ class SomeClassSerializer(StreamSerializer):
         token_provider (hazelcast.security.TokenProvider): Token provider for
             custom authentication (Enterprise feature). Note that token_provider
             setting has priority over credentials settings.
+        use_public_addresses (bool): When set to ``True``, client uses the
```
[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/client.py]: I am also tried to check with grammarly.

looks like these are left regarding `client`:

* **Client's** heartbeat requests are timed out.
* When set to ``True``, **client** statistics collection
* **Client** shuffles the given member list

**Review by srknzl — APPROVED** [review]
I left one last additional comment in one of the previous review comments. Thanks for your patience and changes.

**Comment by yuce** [comment]
@mdumandag Thanks. I agree that `address` would be better. Not sure why the core/Java client property uses `ip`...
